### DISCUSSION TITLE
Isolated Web Apps

### DISCUSSION THREAD

**Discussion by MzHub** [Devographics/surveys#170] [State of HTML Features]
### Specification URL:

https://github.com/WICG/isolated-web-apps

### Any other links?

https://chromestatus.com/feature/5146307550248960

### How stable is this feature?

Future (No implementations yet)

### Type of feature

Other

### What focus area does this feature fall under?

Security & Privacy, Bridging the gap with native

### Details

Isolated Web Apps is related to [Subresource Integrity](https://github.com/Devographics/surveys/discussions/145) and [PWA manifest](https://github.com/Devographics/surveys/discussions/119). It provides a missing security layer that native apps have, which is signed updates to the app. A notable example is [Signal's refusal](https://github.com/WICG/isolated-web-apps/blob/64ee186/README.md?plain=1#L11) to implement a web version of their app due to this missing piece. It would be interesting to see if other developers also avoid building web apps due to this.

**Comment by LeaVerou** [comment]
Is there a code snippet that shows how this works? I could not figure it out from the explainer.

**Reply by MzHub** [reply]
I have to admit I'm not deeply familiar with the explainer, but here are code-adjacent snippets, if it helps at all.

From [isolated-app scheme explainer](https://github.com/WICG/isolated-web-apps/blob/main/Scheme.md):
```
isolated-app://signed-web-bundle-id/path/inside/app.js?some-query#foo
^           ://^                   /^                 ?^         #^
scheme         opaque host          path               query      fragment
```
The proposal is that isolated apps could be delivered as [Web Bundles](https://wpack-wg.github.io/bundled-responses/draft-ietf-wpack-bundled-responses.html), which are "bundled up groups of HTTP responses". I'm not sure on the details further than that.

-----

From [updates manifest explainer](https://github.com/WICG/isolated-web-apps/blob/main/Updates.md#web-application-update-manifest):
```json
{
  "versions": [
    {
      "version": "5.2.17",
      "src": "https://cdn.example.com/app-package-5.2.17.swbn"
    },
    {
      "version": "6.1.13",
      "src": "v6.1.13/package.swbn"
    },
    {
      "version": "7.0.6",
      "src": "v7.0.6/package.swbn"
    }
  ] 
}
```
An example of the proposed new manifest file to handle updates to the isolated app.

-----

From [permissions explainer](https://github.com/WICG/isolated-web-apps/blob/main/Permissions.md):
```json
"permissions_policy": {
   "geolocation": [ "self", "https://map.example.com" ],
   "fullscreen": [ "*" ]
}
```
Explicit permissions similar to how apps work on mobile. To give users more awareness and control.

-----

Sorry for lack of material here, I come at this more from wanting a solution, although I do not fully understand the solution.

For context, there are some efforts at solving this from different direction, for example there is [SecureBookmarks](https://github.com/coins/secure-bookmark), which attempts to solve the "can't trust the server"-issue by placing a subresource integrity hash inside a bookmark. A malicious server can not change the user's bookmark, so it is a safe starting point for the app.

Then there's [Signed Pages](https://github.com/tasn/webext-signed-pages) which attempts to solve this, or at least similar issues, via a browser extension.

**Reply by LeaVerou** [reply]
Thanks! I was more looking for a 1-3 line snippet to include in the question. Maybe an `isolated-app://` URL?

**Reply by MzHub** [reply]
Closest I can find to "how it works in practice right now" is this https://github.com/GoogleChromeLabs/telnet-client/

```sh
npm run start

google-chrome-unstable --enable-features=IsolatedWebApps,IsolatedWebAppDevMode \
                       --install-isolated-web-app-from-url=http://localhost:8080
```

or

```sh
npm run build

google-chrome-unstable --enable-features=IsolatedWebApps,IsolatedWebAppDevMode \
                       --install-isolated-web-app-from-file=$PWD/dist/telnet.swbn
```

**Reply by MzHub** [reply]
I still don't have a suitable code snippet for the survey, but I have a suggestion for the short explanation text, in case there isn't one yet.

This is based on the text from [Isolated Web Apps Chrome Platform Status](https://chromestatus.com/feature/5146307550248960) overview, which I think is good, but a bit long for the survey.

The overview reads:

>Isolated Web Apps (IWAs) are an extension of existing work on PWA installation and Web Packaging that provide stronger protections against server compromise and other tampering that is necessary for developers of security-sensitive applications.
>
>Rather than being hosted on live web servers and fetched over HTTPS, these applications are packaged into Web Bundles, signed by their developer, and distributed to end-users through one or more of the potential methods described in the explainer.

I would suggest for the survey:

>Native-like packaging, permissions and signed updates for PWAs.

or

>Native-like packaging, permissions and secure updates for PWAs.

-----

Maybe you're right that out of not-so-good options the isolated-app url is a best snippet to at least illustrate that they're not server hosted.

```
isolated-app://4tkrnsmftl4ggvvdkfth3piainqragus2qbhf7rlz2a3wo3rh4wqaaic/index.html
```

(I think in reality the `index.html` part probably isn't there, it's just `/` as that is usually the `start_url` in the PWA manifest, but the main file in the IWA demo app is [`index.html`](https://github.com/GoogleChromeLabs/telnet-client/blob/main/src/index.html))

**Reply by LeaVerou** [reply]
This is very useful, thanks!
